src/hotspot/share/gc/shared/partialArrayTaskStepper.inline.hpp

```diff
@@ -0,0 +1,106 @@
+  // because length is a multiple of the chunk size.  Both of those are
+  // relatively rare and handled in the normal course of the iteration, so
+  // not worth doing anything special about here.
+  to->set_length(end);
+
+  // If the initial chunk is the complete array, then don't need any partial
+  // tasks.  Otherwise, start with just one partial task; see new task
+  // calculation in next().
+  Step result = { end, (length > end) ? 1u : 0u };
+  return result;
+}
+
+PartialArrayTaskStepper::Step
+PartialArrayTaskStepper::next(arrayOop from, arrayOop to, int chunk_size) {
+  assert(chunk_size > 0, "precondition");
+
+  // The start of the next task is in the length field of the to-space object.
+  // Atomically increment by the chunk size to claim the associated chunk.
+  // Because we limit the number of enqueued tasks to being no more than the
+  // number of remaining chunks to process, we can use an atomic add for the
+  // claim, rather than a CAS loop.
+  int start = Atomic::fetch_and_add(to->length_addr(),
+                                    chunk_size,
+                                    memory_order_relaxed);
+
+  // The from-space object contains the real length.
+  int length = from->length();
+  assert(start < length, "invariant: start %d, length %d", start, length);
+  assert(((length - start) % chunk_size) == 0,
+         "invariant: start %d, length %d, chunk size %d",
+         start, length, chunk_size);
+
+  // Determine the number of new tasks to create.
+  // Zero-based index for this partial task.  The initial task isn't counted.
+  uint task_num = (start / chunk_size);
+  // Number of tasks left to process, including this one.
+  uint remaining_tasks = (length - start) / chunk_size;
+  assert(remaining_tasks > 0, "invariant");
+  // Compute number of pending tasks, including this one.  The maximum number
+  // of tasks is a function of task_num (N) and _task_fannout (F).
+  //   1    : current task
+  //   N    : number of preceeding tasks
+  //   F*N  : maximum created for preceeding tasks
+  // => F*N - N + 1 : maximum number of tasks
+  // => (F-1)*N + 1
+  assert(_task_limit > 0, "precondition");
+  assert(_task_fannout > 0, "precondition");
+  uint max_pending = (_task_fannout - 1) * task_num + 1;
+
+  // The actual pending may be less than that.  Bound by remaining_tasks to
+  // not overrun.  Also bound by _task_limit to avoid spawning an excessive
+  // number of tasks for a large array.  The +1 is to replace the current
+  // task with a new task when _task_limit limited.  The pending value may
+  // not be what's actually in the queues, because of concurrent task
+  // processing.  That's okay; we just need to determine the correct number
+  // of tasks to add for this task.
+  uint pending = MIN3(max_pending, remaining_tasks, _task_limit);
+  uint ncreate = MIN2(_task_fannout, MIN2(remaining_tasks, _task_limit + 1) - pending);
+  Step result = { start, ncreate };
+  return result;
```

Comment:
Similar to my comment above, what do you think about trying to write some test for this to verify we never get to many "tasks".

Reply:
The asserts in the stepper's next() after the increment of to's length verify that we haven't overrun.  I haven't thought of a way to verify the algorithm doesn't generate too few tasks though.  Well, other than getting crashes because some array elements didn't get processed.  But maybe you mean unit tests?  I will try to write some; I should have done so already.

Reply:
Should have been a bit more clear. I'm talking about unit tests, so if you plan writing some that's great.